### ISSUE TITLE
remote_tmp not honored for async tasks

### ISSUE THREAD

**Issue by IOOOTAlan** [ansible/ansible#28318]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
async

##### ANSIBLE VERSION
Originally tested on v2.3.0:
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan  8 2017, 11:53:02) [GCC 5.3.0]
```
but found it is present also in
```
ansible 2.3.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan  8 2017, 11:53:02) [GCC 5.3.0]
```

##### CONFIGURATION
Using these settings (via ANSIBLE_CONFIG file override):
```ini
[defaults]

# /home dirs are not writable. Use /tmp instead
local_tmp = /tmp
remote_tmp = /tmp
host_key_checking = False

[ssh_connection]

pipelining = True

ssh_args = -o ControlMaster=auto -o ControlPersist=60s -o ConnectTimeout=30
```

##### OS / ENVIRONMENT
Running Ansible from Slackware Linux 14.2.

Managed host OS is Ubuntu 16.04 on ARM architecture.

##### SUMMARY
As mentioned in #10054 for Ansible version 1.8.2, it seems like the `remote_tmp` setting inside the configuration file is not honored by the async module.
I'm trying to perform a set of tasks on remote systems where the root partition is normally mounted in read-only mode (`/home` is inside the root partition), so I need to override the `remote_tmp` setting to make common Ansible tasks work.
When using `async` directives to launch commands in a "fire and forget" fashion (i.e. setting `poll` to 0), Ansible tries to write inside the remote user home directory, regardless of the `remote_tmp` override, failing miserably.

##### STEPS TO REPRODUCE
 1. save the previously reported configuration inside a `ansible-asyncTest.cfg` file
 2. launch the following playbook like: `ANSIBLE_CONFIG=ansible-asyncTest.cfg ansible-playbook -vv ...`

```yaml
---
- hosts: testAsync
  gather_facts: yes
  tasks:
    - name: test_async
      command: "sleep 15"
      async: 9
      poll: 0
```

##### EXPECTED RESULTS
The task should run with no errors

##### ACTUAL RESULTS
The task fails with the following error:
```
fatal: [192.168.0.19]: FAILED! => {"changed": true, "failed": true, "module_stderr": "Shared connection to 192.168.0.19 closed.\r\n", "module_stdout": "{\"msg\": \"could not create: /home/remoteUser/.ansible_async\", \"failed\": 1}\r\
n{\"started\": 1, \"_ansible_suppress_tmpdir_delete\": true, \"finished\": 0, \"results_file\": \"/home/remoteUser/.ansible_async/39239918689.7773\", \"ansible_job_id\": \"39239918689.7773\"}\r\n", "msg": "MODULE FAILURE", "rc": 0}
```

**Comment by nrwahl2** [comment]
Here's where it's being set:
```
[reid@laptop ~/git/ansible/lib/ansible/modules/utilities/logic]$ grep -n 'jobdir =' async_wrapper.py 
220:    jobdir = os.path.expanduser("~/.ansible_async")
```

What I would need to know (core team pls help) is where variables like `remote_tmp` are stored that modules can access. I've been digging and can't seem to find the Great Dictionary. If we can find it, then we should be able to conditionally override the `jobdir` value.

Edit: I'm only finding them available through action plugins importing `playbooks/play_context.py`. Based on some of the function args it looks like async is set up either before or simultaneously with the action plugins.

**Comment by nrwahl2** [comment]
Got it! The changes needed to be made in `plugins/action/__init__.py` and `modules/utilities/logic/async_wrapper.py`. In `plugins/action/__init__.py`, `remote_tmp` is already pulled into a variable called `tmp` through a few steps in `_make_tmp_path`, and is then in scope for the `if wrap_async` block.

(Ignore the host pattern match warnings. Playbook points to localhost but I forgot to pass the inventory file on the command line.)

```
[reid@laptop ~/git]$ cat ansible_async_test.cfg 
[defaults]

# /home dirs are not writable. Use /tmp instead
local_tmp = /tmp
remote_tmp = /tmp
host_key_checking = False

[ssh_connection]

pipelining = True

ssh_args = -o ControlMaster=auto -o ControlPersist=60s -o ConnectTimeout=30
[reid@laptop ~/git]$ cat async-test.yml
---
- hosts: localhost
  gather_facts: yes
  tasks:
    - name: test_async
      command: "sleep 15"
      async: 9
      poll: 0
[reid@laptop ~/git]$ ANSIBLE_CONFIG=ansible_async_test.cfg ansible-playbook -vvv async-test.yml
ansible-playbook 2.4.0 (issue/10054 8597ddf3d8) last updated 2017/08/27 19:30:03 (GMT -500)
  config file = /home/reid/git/ansible_async_test.cfg
  configured module search path = [u'/home/reid/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/reid/git/ansible/lib/ansible
  executable location = /home/reid/git/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Jun 26 2017, 10:20:05) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
Using /home/reid/git/ansible_async_test.cfg as config file
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: Could not match supplied host pattern, ignoring: all

 [WARNING]: provided hosts list is empty, only localhost is available


PLAYBOOK: async-test.yml ******************************************************************************************************************************************************************************************
1 plays in async-test.yml

PLAY [localhost] **************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
Using module file /home/reid/git/ansible/lib/ansible/modules/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: reid
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /tmp/ansible-tmp-1503889047.15-155175780734799 `" && echo ansible-tmp-1503889047.15-155175780734799="` echo /tmp/ansible-tmp-1503889047.15-155175780734799 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmp0Au9ra TO /tmp/ansible-tmp-1503889047.15-155175780734799/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /tmp/ansible-tmp-1503889047.15-155175780734799/ /tmp/ansible-tmp-1503889047.15-155175780734799/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /tmp/ansible-tmp-1503889047.15-155175780734799/setup.py; rm -rf "/tmp/ansible-tmp-1503889047.15-155175780734799/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [test_async] *************************************************************************************************************************************************************************************************
task path: /home/reid/git/async-test.yml:5
Using module file /home/reid/git/ansible/lib/ansible/modules/commands/command.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: reid
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /tmp/ansible-tmp-1503889048.69-40899558175828 `" && echo ansible-tmp-1503889048.69-40899558175828="` echo /tmp/ansible-tmp-1503889048.69-40899558175828 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmp0tKPOi TO /tmp/ansible-tmp-1503889048.69-40899558175828/command.py
<127.0.0.1> PUT /tmp/tmpCcQsYE TO /tmp/ansible-tmp-1503889048.69-40899558175828/async_wrapper.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /tmp/ansible-tmp-1503889048.69-40899558175828/ /tmp/ansible-tmp-1503889048.69-40899558175828/command.py /tmp/ansible-tmp-1503889048.69-40899558175828/async_wrapper.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /tmp/ansible-tmp-1503889048.69-40899558175828/async_wrapper.py 176746365995 9 /tmp/ansible-tmp-1503889048.69-40899558175828/command.py _ /tmp/ansible-tmp-1503889048.69-40899558175828/ && sleep 0'
changed: [localhost] => {
    "ansible_job_id": "176746365995.12262", 
    "changed": true, 
    "failed": false, 
    "finished": 0, 
    "results_file": "/tmp/ansible-tmp-1503889048.69-40899558175828/176746365995.12262", 
    "started": 1
}
META: ran handlers
META: ran handlers

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   
```


The question now will be, do we want to make this the default behavior (which to me would actually make sense, as much as I love promoting configurability), or make this a separate override option? The way it's set up with my changes, it uses `remote_tmp` (unless something passes a `tmp` argument to `plugins.action._execute_module`, which should only happen in specialized cases with someone who knows what they're doing).

I'm about to submit a PR.

**Comment by nrwahl2** [comment]
@IOOOTAlan @jimi-c @nitzmahone : Created PR #28716

**Comment by IOOOTAlan** [comment]
Thanks a lot @nrwahl2!

**Comment by nrwahl2** [comment]
You're quite welcome! Though I didn't realize this was going to require me to mess with Powershell. Grr, Windows....

**Comment by alikins** [comment]
See also https://github.com/ansible/ansible/issues/31022

**Comment by bcoca** [comment]
fixed via #31677, take note that async tasks still use an independant 'async status dir' which is also now configurable.
